### DISCUSSION TITLE
CMIP branded-variable proposal

### DISCUSSION THREAD

**Discussion by durack1** [WCRP-CMIP/WGCM_Infrastructure_Panel#36] [General]
This discussion pulls an email thread out of inboxes

**Comment by durack1** [comment]
From: "Taylor, Karl E."
Date: Sunday, November 13, 2022 at 8:35 AM
To: "m.n.juckes@", "Mizielinski, Matthew", "Durack, Paul J."
Subject: Re: CMIP branded-variable proposal

Hi Martin, Paul, and Matt (and relying again on Matt to deliver the mail to Martin for me)

I’ve responded to Martin’s comments below.  I don’t think we’re too far apart in our thinking and look forward to any further comments/suggestions you have.

Best regards,
Karl

(A) MJ: First on "realms": we dropped this from the initial call for data request submissions in CMIP6 and got very strong feedback that the community needed it, so I doubt they will be happy with losing it now. I expect we will have to define the realm attribute, whether or not it is in the branded name. Variable names are not objective, so I don't see that the absence of a fully objective method of assigning realm is a decisive argument here. As you note later, variables will have realms assigned. The question here is whether we use it or not.

KET responds:  Regarding “the community needed it”, do you recall for what purpose? Does “realm” need to be in the branded name to fulfill the community’s needs?

KET responds: Regarding “doubt they will be happy with losing it”, I don’t propose doing away with the “realm” (or “modeling_realm”) attribute. I suggest removing it from the variable label.  For many variables, I think “realm” should be assigned multiple values (as in CMIP6).  The “realm” attribute” could be useful for multiple purposes including: 1) identifying variables that are produced by a particular model component (atmosphere, land, sea ice, etc.), and 2) identifying variables that might be of special interest to a particular “community” (and used in filtering searches).

MJ: It is clear that the realm is often superfluous for identification. Tables such as Omon and CFmon combine variables from multiple realms which are uniquely identified by the name and table. The same is true of the abbreviated area types which you are proposing. Often superfluous, but one or the other of these categories is needed to have a robust system which allows flexibility. 

It would be helpful, I think, to have a clear statement of what the branding is aiming to achieve. Such as: 
1.	A transparent and systematic approach to variable naming;
2.	Names which can be readily parsed by machines and by humans to a set of important characteristics;
3.	A system with sufficient flexibility to accommodate significant expansion in the number of names with revision of existing names;
4.	Names which can be used within multiple contexts (file names, names in code, databases, text);
KET responds: Regarding the above, I think in addition we should strive to limit the root names to one per standard_name whenever possible.  Thus, the variable name could serve as an abbreviation for the standard_name.  We know we can’t always achieve this though.
MJ: Using the "where" element of the cell methods does give a systematic approach to identifying an area type suffix. However, as area types have clear links to realms, it also gives a systematic approach to identify a realm. I think realms win on intelligibility to humans and that requiring a dictionary to translate from area type abbreviations to area types will be a significant overhead for human and software parsers.

KET responds: Regarding “area types have clear links to realms”, it’s true that you could assign each area type to one of a few realms.  For example, both of the area types “sea”and “ice_free_sea” could be assigned to a single realm=”sea”, but then you would need two different variable roots to distinguish surface_temperature “where sea” from surface_temperature “where ice_free_sea”.  I think it is cleaner to have a single variable root (“ts”) identify surface_temperature no matter where, and then use area type branding to distinguish between “sea” and ice_free_sea”.

KET responds: Regarding “it [the ‘where’ element] also gives a systematic approach to identify a realm”, I originally thought along those lines, but when I tried to implement such an approach I failed.  I think the “realm”, as we’ve defined it in the past has been assigned considering two primary aspects: what model component was responsible for calculating the variable, and what users might associate with the variable (in filtering searches).  The first decision to make is how many realms should be defined?  For CMIP5 and CMIP6 we had eight: atmos, ocean, land, landIce, seaIce, aerosol, atmosChem, and ocnBgchem.  One could argue for a finer granularity, for example, “land” split between “soil” and “vegetation” or simply detaching “landBgchem” from “land”.  However these categories are defined, determination of the realm(s) associated with any variable can be difficult.  Any variable defining a “flux” at an interface between two reservoirs (e.g., between atmosphere and ocean) might be assigned with equal justification to either of the two.  We generally assigned these variables to both categories, but our determination of which category should be primary was somewhat arbitrary. Similarly, certain variables are clearly of interest to multiple realms. Methane concentration in the atmosphere could be included in both atmosChem (because it impacts chemical reactions) but also “atmos” (because it impacts radiative transfer).  Which should be the primary realm for methane concentration?  It was questions like these that led me to the much simpler and objective approach of defining the areaLabelDD for the purposes of uniquely identifying variables.  We can define the realm for other purposes.

(B) MJ: I was suggesting "s", rather than "zs", but I think your comments apply equally. I wouldn't want to use either "z0" or "s" for the base of the thermocline or tropopause. Both of these are surfaces defined by a transition in stratification. We could either look for a letter or letters to stand for a stratification vertical coordinate, or use specific identifies such as "tp" and "tc". 

KET responds: Regarding the above, I was attempting to do the simplest thing here, so that the verticalLabelDD could be automatically generated by just determining if a vertical coordinate was or was not defined for a variable.  If no vertical coordinate were defined, then “z0” would be assigned (but I’m open to suggestion for a different text string).  [The “z0” was supposed to be analgous to z1, z2, z3, etc. which indicate reporting at 1, 2, and 3 depths in the vertical, respectively; z0 would indicate that there was no explicitly defined depth (i.e., no vertical coordinate).]   Distinguishing among all the individual types of variables that don’t have vertical coordinates leads to an algorithm that is very (hopelessly?) complicated.  I note among the various types of variables we need to account for are: fluxes at the top of the atmosphere, quantities at the interface between the atmosphere and the medium below (i.e., the ”surface”), quantities at the interface between the ocean surface and sea ice, quantities at the interface between the land ice and the medium below it (land or sea), quantities at the interface between snow and the medium below, the “tropopause”, the “base of the thermocline”, at cloud top, at cloud base, at mean sea level, integrals through a column of the atmosphere, ocean, sea ice, and many, many more.  (Note that not all of these are surfaces coincide with “a transition in stratification”.)  Each of these would need a different verticalLabelDD, which no one would likely be able to remember.  I think I’d like to stick with “z0”, or some other generic label, that simply means “no vertical coordinate”.   We might, for example, use “x” to indicate this, rather than “z0”, because “x” might be interpreted and remembered as “vertical coordinate excluded”.

MJ: Is "h010" easier to parse than "10m"? I'm not sure. I suppose your proposal can be parsed with a shorter piece of code, but I think the difference is marginal. For humans, however, there is a big difference in legibility.

KET responds: Regarding the options for representing a field reported on a single coordinate surface in the vertical, I was concerned that an ocean variable requested on a particular density surface would need a rather long label: for example, the density surface “1026 kg m-3” could be written “1026kgm-3”, but the hyphen might be problematic and the normal unit for this is "sigma-t” (density in kg m-3 - 1000 kg m-3), which I would have trouble abbreviating.  Thus, I was led to standardizing the unit for each type of vertical coordinate, but omitting the unit from the label.  Thus, instead of 1026kgm-3, we would have rho26.  Perhaps that concern should not dominate our thinking here, though.  By including the units, we might avoid having decimal points represented by “p” (under my scheme) and there would be no need for the “prefix” on the verticalLabelDD identifying what type of vertical coordinate we have (e.g, instead of p0500, we would have “500hpa” because “hpa” would imply a pressure coordinate).  I note that if we adopt this alternative, “1 meter depth below the surface” and “1 meter height above the surface” would have the same label: “1m”.  Could there be some case where this ambiguity would be a problem?  If not, I think I now favor your approach here.

(C) MJ: OK, I think the 6 classes would be cleaner.

(D) MJ: OK .. perhaps we could offer two versions of the area type vocabulary for consultation: one short version and one expanded for legibility. I hope we can also offer an option of using the realms.

(E) MJ: OK, that sounds good,

From: Martin Juckes - STFC UKRI
Date: Thursday, November 10, 2022 at 10:21 AM
To: "Mizielinski, Matthew", "Taylor, Karl E."
Cc: "Durack, Paul J.", "Taylor, Karl E."
Subject: Re: CMIP branded-variable proposal

Hi Karl,

(A) First on "realms": we dropped this from the initial call for data request submissions in CMIP6 and got very strong feedback that the community needed it, so I doubt they will be happy with losing it now. I expect we will have to define the realm attribute, whether or not it is in the branded name. Variable names are not objective, so I don't see that the absence of a fully objective method of assigning realm is a decisive argument here. As you note later, variables will have realms assigned. The question here is whether we use it or not.

It is clear that the realm is often superfluous for identification. Tables such as Omon and CFmon combine variables from multiple realms which are uniquely identified by the name and table. The same is true of the abbreviated area types which you are proposing. Often superfluous, but one or the other of these categories is needed to have a robust system which allows flexibility. 

It would be helpful, I think, to have a clear statement of what the branding is aiming to achieve. Such as:
1.	A transparent and systematic approach to variable naming;
2.	Names which can be readily parsed by machines and by humans to a set of important characteristics;
3.	A system with sufficient flexibility to accommodate significant expansion in the number of names with revision of existing names;
4.	Names which can be used within multiple contexts (file names, names in code, databases, text);

Using the "where" element of the cell methods does give a systematic approach to identifying an area type suffix. However, as area types have clear links to realms, it also gives a systematic approach to identify a realm. I think realms win on intelligibility to humans and that requiring a dictionary to translate from area type abbreviations to area types will be a significant overhead for human and software parsers.

(B) I was suggesting "s", rather than "zs", but I think your comments apply equally. I wouldn't want to use either "z0" or "s" for the base of the thermocline or tropopause. Both of these are surfaces defined by a transistion in stratification. We could either look for a letter or letters to stand for a stratification vertical coordinate, or use specific identifies such as "tp" and "tc". 

Is "h010" easier to parse than "10m"? I'm not sure. I suppose your proposal can be parsed with a shorter piece of code, but I think the difference is marginal. For humans, however, there is a big difference in legibility.

(C) OK, I think the 6 classes would be cleaner.

(D) OK .. perhaps we could offer two versions of the area type vocabulary for consultation: one short version and one expanded for legibility. I hope we can also offer an option of using the realms.

(E) OK, that sounds good,

regards,
Martin
 
From: Taylor, Karl E.
Sent: 02 November 2022 01:03
To: m.n.juckes@
Cc: Durack, Paul J.; Mizielinski, Matthew 
Subject: Re: CMIP branded-variable proposal
 
This email was received from an external source.   Always check sender details, links & attachments.
Hi Martin,  (sorry to impose Matt, but could you please forward again to Martin?)
 
Thanks for giving this your careful attention.  I’ve responded, at least partially, below.  As you can guess, I’ve devoted considerable thinking (and way too much time) devising a reasonable easy-to-apply methodology to uniquely define the CMIP6 variables.  Even so, you have suggested a few modifications that are certainly worth considering.  I am hopeful that with further iteration we can reach a consensus on the best way forward with a proposal that the modeling groups and users might also seriously consider and provide helpful feedback. 
 
Best regards,
Karl
 
From: Martin Juckes - STFC UKRI
Date: Tuesday, November 1, 2022 at 2:27 AM
To: "Taylor, Karl E."
Cc: "Durack, Paul J." , "Mizielinski, Matthew"
Subject: Re: CMIP branded-variable proposal
 
Hello Karl,
 
Thanks for the clarifications. I'm relieved that we are not necessarily considering renaming large numbers of variables.
 
I was intrigued by your statement that using "<variable>-<realm>-<frequency>" would require "many more" variables to be renamed and took a closer look. By my reckoning there would be around 100 variables that needed to be adjusted. After looking a bit closer I've come to agree that something closer to your proposal would be better. Especially because it would, as you have said, give more flexibility to accommodate an expected influx of new variables.
 
Among those 100 there are a few which appear to be there because of errors/inconsistencies in the specification of the primary realm in the data request: either variables masked by land which are specified as "atmos" or variables masked by snow which are specified as "landIce". The convention needs to be clarified here, but I believe it should be that masking a global variable in a way which restricts the data to be within a spatially defined realm (i.e. land, ocean, landice or seaice) then the primary realm should be specified correspondingly. 
 
KET: >>>>> see comments below item (A) below.
 
MJ:
(A) I'd like to propose an alternative which, I believe, also requires no name changes (though I agree that some may be desirable) and has some potential advantages: <variable>-<realm>-<frequency>-<time_label>-<horizontal_label>-<vertical_label>. This uses "realm" instead of the "areaLabelDD". 
 
Advantages of "realm" over "areaLabelDD": 
     (1) the "realm" terms are familiar to users and the list is short and stable, the list of area types is long and likely to expand. People will not generally understand the cryptic abbreviations. 
     (2) Preserves a clear link with existing system: the <time_label>-<horizonatal_label>-<vertical_label> element would correspond roughly to some of the sub-tables in the larger CMIP5/6 MIP tables, but with more systematic naming. The tables with the longest heritage and probably the highest usage, such as "Amon", "Omon", will then be mapped into a well-defined part of the new system, and it will be easy to reconstruct them directly from the vocabularies used to define variables. 
     (3) The "realm" corresponds to scientific communities and is hence provides a useful means of quickly identifying a group of variables which may be of interest to a given community.  
 
KET response: I spent many days trying to find an objective, justifiable, and easy way to assign a primary realm and possibly some secondary realms to each variable.  As you note, in the past “realms” helped group together variables that might be of particular interest to certain communities.  But of course some variables are of interest to multiple communities, so there must be some arbitrariness in deciding which realm a variable belongs to – for example, surface temperature is of interest to those focusing on the atmosphere, but also those studying any of the surface realms).  Or consider aerosol variables which could be considered to be in the “atmospheric physics” realm (affecting radiative transfer and cloud physics) and the “atmospheric chemistry” realm (affecting chemical reactions in the atmosphere).  It can be difficult to decide which realm should be primary.  
 
KET response cont.: After failing to find an objective method by which a variable’s primary realm could invariably be determined, I realized that the “realm” is more often than not superfluous for unique identification of a variable.  The realm is almost always determined by the standard name (e.g., air_temperature), which is already implied by the root name (already included in the branded variable label).  So it is rarely necessary to additionally include the realm as a separate part of the branding suffix.  What is needed for unique identification in some cases, however, is whether a variable has been sampled at all grid cells or only a subset of cells characterized by an area type that is specified in the “where” directive of a cell method.  So it is the area type that is needed for unique identification, and that is why I have suggested using the area_types in the “where” directive for uniquely identifying variables.  
 
KET response cont.: That being said, I think it is essential that we continue to define for *other* purposes the realm or realms a variable belongs to.  I think, for example, that a data request might want to group variables by realm (as we have in the past in distinguishing between Amon and Omon variables).  Equally important, I think ESGF should include “realm” as a search facet so that users can filter the long list of variables to a subset that might be of most interest to them.  I have not set down the rules for specifying the realm associated with each variable because I don’t think realm is needed to uniquely identify them.  We will certainly need to do that in defining the full set of variable attributes, and in the spreadsheet you have access to (for CMIP6 variables), some of the hidden columns include realm designations determined through complex algorithms taking as input the standard_name and the cell_methods.
 
MJ:
(B) For identifying vertical levels, plev19 etc, it seems clear that there needs to be a short identifier somewhere and having a clearly defined vocabulary will make this clearer. 
   (1) For pressure, you propose "p<nn>" for "nn" levels and "p0<xx>" for a level at xx hPa, and a similar approach for height, so that the qualifier "10m" is translated into "h010". I would prefer to stay with "10m" because, I think, this will be more readily interpreted by users (also one character shorter). By analogy, that would imply using, for instance "100hpa" rather than "p0100". Again, I think this is preferable on grounds of clarity, though it is one character longer. Soil depths would be in "cm" or "m" as appropriate.
  (2) For surface variables, I would preserve "s" rather than introduce "z0".
 
KET response: I think that your approach is probably easier for humans to understand, and my approach might be slightly easier for a computer to parse:  if the first numerical digit encountered is 0, then the field is not a function of the vertical coordinate.  The problem of substituting “zs” for “z0” is that we need to accommodate variables that represent a quasi-horizontal slice at say the “tropopause” or the “base of the thermocline”, which cannot be defined by a single coordinate value and obviously are not “surface” variables.  Maybe you have an idea on how to represent these differently?
 
MJ:
(C) Temporal label: can we keep this down to 5 classes: fixed, point, interval, annual climatology, diurnal climatology? I feel that this would be cleaner because these 5 are essentially about the sampling of data, while the distinction between statistics, average and sum is about what you do with the sampled data. The distinction between these three categories has always been in the variable names, so there is no requirement for a change here. When a variable is defined as a statistic this should be in the standard name, so we will, if we follow the current de facto convention that we don't have multiple standard names mapping onto a single variable label, always have to reflect the nature of the statistic in the variable name.
 
KET response: This, as you imply, is somewhat a judgement call.  Those preparing data, I think, might like to know which variables represent time-means as opposed to some other more complex statistical measure like the maximum temperature within an interval, which generally would require a different sort of processing algorithm.  That was the rationale for defining tavg, distinct from “tsum” or “tstat.  That being said, “tsum” is only needed for 3 of the CMIP6 variables, so perhaps it should be eliminated and be included in the “tstat” category.  Thus, we would end up with 6 classes, with your “interval” expanded into “tavg” and “tstat” (or some other appropriate name).  
 
MJ:
(D) Renaming: 
   (1) I think we agree that renaming "tas", "uas" etc to remove the redundant "s" is not a good idea, and we should only rename where we have a clear benefit and assessment of the potential downsides.
 
KET response: Yes we agree on this.  We might also agree that moving forward we should not necessarily adopt all the practices of the past.  We might construct the root names of new variables (without any established history) differently from the past.
 
MJ:
   (2) We have variable like "ua" for multilevel data. If we rename single level fields, "ua200" etc, perhaps it would be worth keeping them separate from the multilevel fields by naming them something like "uax"? Naming them "ua" would be sufficient, but I have reservations about using the same variable name for single level and multi-level data. Leaving them as "ua200" etc is also possible, of course, but the repetition of "200" is a little untidy. 
 
KET response: “uax” is an interesting idea.  Will think about it more.
 
MJ:
  (3) If we keep "areaLabelDD", rather than using my suggestion of "realm", there would be a lot of repetition in names, with, in some cases, two distinct abbreviations of an area type (e.g. cSoilGrass-mon-tavg-l-hxy-ng in which "grass" and "ng" are abbreviations of "natural_grasses"). Cutting out "Grass" makes the name more cryptic than it was, so I don't see a clean solution here .. reinforcing my preference for using <realm> and keeping the names of these variables unchanged.
 
KET response: Yes, the root name would contain information that would be redundant with its areaLabelDD.  There are I think about 8 variable roots that currently describe the main properties of surface vegetation (veg. height, veg. carbon content, CO2 exchanges, gross primary productivity, net primary productivity, litter carbon content, soil carbon content, and respiration).  For each of the variables in CMIP6, we define multiple root names to specify which vegetation type is being characterized (including grass, shrub, tree, crop, pasture, leaf, wood, c3veg, c4veg).  In CMIP6 we had no “areaLabelDD”, and so had to include the type of vegetation in the root name.  If we had defined an areaLabelDD, we could have reduced the number of root names pertaining to vegetation to about ¼ the original number.  Moreover, users (and perhaps more importantly software) would be able to determine the vegetation type by explicit rules, rather than interpreting the root name.   
 
KET response cont.: I do agree that a 2-letter areaLabelDD’s like “ng” may not be ideal.  Perhaps these labels should be made longer (e.g., replace “ng” with “grass”).   I also think we need to decide whether to modify the root names of past names that included the area type, or leave them unmodified, but recommend that future variables leave off the redundant information in the root because it will be recorded by areaLabelDD.
 
MJ:
(E) Consultation: I don't think you picked up on my point about consultation, but I feel that some of the questions raised above might be better resolved by asking some experienced users such as, at least, the leaders of MIPs that have made major contributions to the CMIP variable vocabulary over multiple cycles. The acceptance by the community is likely to be better if we can offer options (as long as it is a limited number of well thought-through options),
 
KET response: I agree that once the WIP has a solid proposal (perhaps with some different options also on the table), we should seek input from others.
Best,
Karl
 
regards,
Martin
 
From: Taylor, Karl E.
Sent: 31 October 2022 15:49
To: Juckes, Martin; Mizielinski, Matthew 
Cc: Durack, Paul J.
Subject: Re: CMIP branded-variable proposal
 
This email was received from an external source.   Always check sender details, links & attachments.
Dear Martin,  (Matt, please forward to Martin).
 
Thanks for reading the document so promptly and getting back to me.
 
I agree that changing the root names of variables that have been in use for multiple phases of CMIP would be disruptive, and I think some of these at least should not, in fact, be changed.  (More radically, I would be happy to retain *all* the CMIP6 root names unchanged if that would make the proposal acceptable.)  In the draft spread sheet I was showing what *could* be done under the new branded naming convention, not that we would necessarily want to make all such changes.
 
I also, wasn’t suggesting that we do away with tables.  I was suggesting that the tables be constructed based on criteria independent of any use in uniquely identifying variables.  So I don’t see that constructing unique names in a different way will impact the ease (or difficulty) MIPs have in constructing their data requests.
 
I would note that whether the identifying text strings used to uniquely define variables are included in the root or in a suffix should not be a show-stopper.  If we are to move away from table names as a suffix, the identifying information the table names have conveyed in CMIP6 (regarding what a variable represents) must be moved either to the root or the suffix.  This requirement means we must define labels to represent at least the frequency, the temporal sampling (to distinguish, for example, between “point” and time-mean variables), the vertical sampling (to distinguish, for example, between model-level and pressure-level variables), the “horizontal” grid (to distinguish, for example, between 2-d gridded data and zonal means), and the area restriction (to distinguish between a variable reported only over ice vs. one reported over land).   Note that a branded variable consisting of only the root plus two modifiers (like “tas-atmos-mon”) would be insufficient to accommodate the CMIP6 variables unless many more new (and disruptive) root names were defined.
 
Under my proposal, I have included all 5 of the short text strings modifiers in the suffix (choosing not to disrupt many of the current root names).  With these labels defined, I think that moving forward there will be less of a need to define new root names.  To reiterate, we don’t have to modify existing root names to implement this new approach.  Some existing root names *could* be simplified and shortened if the consensus is that doing so wouldn’t be too disruptive.  For example ta700 could become “ta” since the pressure level would be recorded by the verticalLabel as “p0700”.  But we might choose not to do this.  In the future if a project wanted to collect ta at say 20 hpa and another project wanted to collect ta at say 950 hPa, we might recommend that these *new* variables be assigned the root name “ta” and let the suffix alone define the pressure level (“p020” and “p0950”, respectively).  
 
If we proceeded in this way, I think users will come to appreciate that the new approach is scalable and easy to understand, and I think they will be happy not to memorize too many root names, which may to some seem obscure and difficult to decipher because these roots have been constructed by individuals unaware of any general rules or conventions for making them understandable.  
 
As I noted in the proposal, for this to work smoothly with the past, two dictionaries will need to be created: one dictionary giving the original CMIP6 root name + table name and the corresponding new branded variable name; the other dictionary providing the same information but with the new branded name first. With this, existing codes could easily be modified to handle data stored with the new variable identifiers just as they handled these variables in CMIP6.  (Note that similar tables could be constructed for CMIP5 and possibly CMIP3.)
 
I’m looking forward to modifying the proposal as needed to make it acceptable.
 
Best regards,
Karl

From: Martin Juckes - STFC UKRI
Date: Monday, October 31, 2022 at 2:10 AM
To: "Mizielinski, Matthew", "Taylor, Karl E."
Cc: "Durack, Paul J."
Subject: Re: CMIP branded-variable proposal
 
Dear Karl,
 
I appreciate the effort that has gone into this, but I am concerned that the scope has expanded substantially since we started discussing a rationalisation of the file naming convention. In those initial discussions the view was that variable names would only need to be changed for a small number of variables. The initial plan was for a limited change between the CMIP cycles, what is now proposed looks like a disruptive change during the early stages of CMIP7. I don't think I can support it as it stands.
 
The document does not address the prime reason for the proliferation of tables in CMIP6, which was an insistance of some MIPs of using existing tables for defining requirements which conflicted with a WGCM requirement that new vaiables from other MIPs shouldn't be added in to general requests automatically. We have, I hope, a clearer procedure now and, with CMIP IPO support, a better chance of avoiding communications going astray or being ignored. 
 
Still, I do agree that it would be a good idea to revise what has become a rather opaque system and removing the MIP table labels from the file names is a good approach. 
 
The current proposal appears to have renamed 301 out of 1313 root variable names. Some of these, such as "tas", "uas", etc, have been used by tens if not hundreds of thousands of scientists for years. For most of these people, who will have little interest in the details of the naming system, the change from "tas_Amon" to "ta-mon-tavg-h02-hxy-x" is likely to be confusing. Something like "tas-atmos-mon" would be more transparent and achieve the aim of doing away with the MIP tables. I suspect many thousands of scientists have many of the existing codes used repeatedly in data analysis and visualisation code and will find dealing with a transition very awkward. Sticking to a 3-part categorisation would mean having to use, as now, suffixes to distinguish between different versions of, for instance, monthly air temperature, but the approach used for commonly used variables has been in place for many years and does not need fixing.
 
I'm also concerned that there has been very little consultation. A change from the names used in the past (when was "tas" introduced?) is going to require coding changes for thousands of users. I don't think we can introduce changes on this scale without consulting. 
 
It looks as if a desire to rationalise the way that, for instance, 6hrPlev fields are treated has influenced the design. This does not look proportionate to me: changes to really widely used variables which will inconvenience tens, perhaps hundreds of thousands of users being made for the sake of some abstract ideas about semantic structure of file names.
 
The section on cell methods appears out of place. I also find it confusing that is presented in terms of adding over a discretised field but it is mostly used, in the CMIP request, to describe the intended sub-grid processing for which there is not generally an explicit discretisation.
 
I do support the goal of retiring MIP tables from the file names, but I'm afraid I can't support this approach,
 
regards,
Martin
 
From: Taylor, Karl E
Sent: Saturday, October 29, 2022 1:23:52 AM
To: Durack, Paul J.; Mizielinski, Matthew; Juckes, Martin
Subject: CMIP branded-variable proposal 
 
This email was received from an external source.   Always check sender details, links & attachments.
28 October 2022  
 
Dear Paul, Matt, and Martin,  (Matt, please forward to Martin for me!)  
 
(Let me know if the 3 mb attachment doesn't arrive)
 
I am seeking your support for my now completed and comprehensive proposal for defining unique labels for CMIP variables.  These labels would replace the root+table labeling we've used in the past (labels like "ta_Amon").  The new labeling has been proposed because the old method led to a proliferation of tables, and also a larger than necessary number of distinct variable root names, which are often quite long (20% of the root names exceed 10 characters).  My feeling is that continuing any longer with this increasingly unmanageable approach will become increasingly difficult because it will not easily scale to meet future demands. 
 
The root names under the current system must distinguish variables that differ only in how the physical quantity is reported (e.g., both msftyrhompa and msftyzmpa represent the same physical variable, but one is reported on density surfaces while the other is reported on model levels).  The new labeling allows both variables to share the same root and the distinction between them is made in the suffix, which in general specifies how the variable has been sampled.   
 
The new labels will facilitate community suggestions for new additions by clearly showing which variables have proved useful in the past.  I think it also will prevent newcomers from wasting our time with suggestions of variables that already exist.  As outlined in the google doc (https://docs.google.com/document/d/1FHqbU2qikt92mApcaEgYY-10O2_oJrTWK_pO4omwpo8), which you have already helped in preparing, I suggest we host all the variables that have been found useful in a single tables.  The variables would be arranged in order of their standard_names. They might be rendered in several forms, including perhaps a spreadsheet like the one attached (but considerably simplified).   
 
A MIP's data request would be constructed by drawing on a subset of the variables in the master list.  Note that defining this master list and its branded variables doesn't preclude a MIP from defining aliases for the branded variable labels and organizing the variables in tables (perhaps akin to the tables defined for CMIP6).  It's only at the overarching foundational level that the branded variable labels must be maintained.  
 
To test the feasibility of the branded variable approach, I have applied it retrospectively to the CMIP6 variables.  The excel spreadsheet attached displays in column BG the proposed branded variable labels.  In addition I propose in column CP new "long_names" for each of the variables, which can for the most part be constructed using an algorithm that extracts information from attributes that are defined for each of the variables.  These long names fully distinguish one branded variable from the next in text that is easily understood by scientists.
 
Note that here we are only interested in the first of the 3 spreadsheets in the workbook (namely, the one named "CMIP6_branded_variables").  You'll notice that most of the columns in the spreadsheet have been hidden. The hidden columns contain information not directly relevant to defining the branded variable labels.  
 
Any procedure that prepares or accesses variables in the CMIP archive would undoubtedly be affected by the transition from root+table labels to branded variable labels, but there will be a one-to-one correspondence between them, so a simple dictionary could be used to translate from one to the other.
 
I am convinced that as a foundation for data storage of MIP model output, the approach I've proposed is far superior to the present one, but we will need to make sure the modeling groups and users understand why we are doing it so they won't reject the changes (which don't need to be overly disruptive) just because they are not exactly the same in the past.  I think progress requires we work toward that goal now, and I think your endorsement and support will be essential.
 
I hope you will soon read the google doc describing the approach and perhaps spend a little time reviewing the spreadsheet.  I look forward to your suggestions on how to proceed and if you would like, I would be happy to provide an overview remotely at your convenience.

best regards,
Karl

**Comment by taylor13** [comment]
In applying the proposed approach to constructing branded variables for CMIP6 variables (see [google doc](https://docs.google.com/document/d/1FHqbU2qikt92mApcaEgYY-10O2_oJrTWK_pO4omwpo8/)), I have also generated human-readable full descriptions of each variable that can be generated based on the same information relied on by branded variables.  90% of the descriptions are generated by code without human intervention.  For the remaining 10%, short modifying descriptors are needed.  The results of this exercise could in the future define the "long_name", and these names can be reviewed in column CO of [this spreadsheet](https://docs.google.com/spreadsheets/d/1_y7OnhSvWc_Cjwnf8dSZcvuVC9BQuSQTPYRWYM1Q9Oc/).   The short modifiers supplied by me (for 10% of the variables) are found in column CN. 

Note that the primary purpose of this spreadsheet was to define the branded variable labels, which are given in column BG.

**Comment by taylor13** [comment]
In response to an offline query, perhaps my verbose earlier postings obscured two points:

1. My proposal for branded variable labels (which involve defining an areaLabelDD) should be viewed independently of any proposal to also associate a "realm" or "realms" with each variable.  In the past, realm was loosely used in conjunction with frequency to define tables which served to uniquely define variables.  If we adopt branded variable labels (as proposed), the realm will be unneeded for that purpose, but realms could be defined for other purposes.  If they are needed, someone will have to come up with guidelines for assigning the realms, but that doesn't have to part of this discussion.
3. In the spreadsheet mentioned above, I have taken the liberty to show how we could have assigned variable roots and branding suffixes to CMIP6 variables under the proposed new branding rules.  If we had followed these rules, the number of distinct variable roots needed to uniquely define the variables in CMIP6 would have been reduced by 211 (from 1282 to 1071).  The number of unique standard_names for CMIP6 variables is 929, so 1071 is only 15% larger than the number of standard_names (as opposed to 38% larger).  Under the original naming of CMIP variables, different roots were needed to distinguish such things as the sub-region where the variable applied.  This led, for example, to water_evapotranspiration_flux being variously named "evspsbl", "sidmassevapsubl", "evs", and "esn" to distinguish among the evapotranspiration over all surfaces, sea ice, ice-free sea, and snow.  Under branding, the differences in variables is indicated in a standardized way (that can be machine and human interpretable) using the areaLabelDD to distinguish among the different surfaces, and a single root can be used for water_evapotranspiration_flux.  What needs to be decided in preparation for CMIP7 is whether multiple roots used in the past to represent the same physical quantity (as defined by a standard_name) should be replaced by a single root (in cases when differences are not required for uniqueness).  In some cases where the original root has existed through many phases of CMIP and AMIP, we likely will decide to retain the original name for the convenience of seasoned users.   I think we should seek input from a wider group than the WIP before making a decision to replace existing roots, but I think *some* roots should be replaced.  In defining new variables for CMIP7, we should avoid introducing new roots when they are unneeded.
